05/implement-network-delay.instructor.md

```diff
@@ -12,36 +56,77 @@ def network_delay_time(times, n, source):
     time_needed[source-1] = 0
     # initialize min heap for traversal
     # priority 0 to travel to the source node
-    heap = [[0, source]] 
+    heap = [(0, source)] 
 
     visited = set() # used to record all visited nodes
-    visited.add(source)
 
     # while there are nodes in the heap
     while heap:
         # pop off the closest node
         time, curr_node = heappop(heap)
+        # if the current node has already been visited, continue
+        if curr_node in visited:
+            continue
+        # add the current node to the list of visited nodes
+        visited.add(curr_node)
         # traverse through the current node's neighbors
         for neighbor, neighbor_time in graph[curr_node]:
+            # if the neighbor has already been visited, continue
+            if neighbor in visited:
+                continue
             # calculate the total time to travel to the neighbor
             total_time = time + neighbor_time
             # if the total time is less than the previous time stored to travel to the neighbor
             if total_time < time_needed[neighbor - 1]:
                 # store the total time as the time needed to travel to the neighbor
                 time_needed[neighbor - 1] = total_time
-                # add the node to the list of visited nodes
-                visited.add(neighbor)
                 # push onto heap to be visited later
-                heappush(heap, [total_time, neighbor])
+                heappush(heap, (total_time, neighbor))
 
     # return the max time in time_needed list if all of the nodes have been visited, otherwise return -1
     return max(time_needed) if len(visited) == n else -1
 ```
+
+As written, this approach includes a few checks to help tamp down the growth of the heap. For example, if a node has already been visited, we can skip it. If the total time to reach a node is greater than the time we already have stored, we can skip it. These checks are not strictly necessary, but they do help to reduce the size of the heap and the number of operations we need to perform.
+
+However, without the `decrease_key` operation, we cannot update the priority of a node that is already in the heap. This means that we will have to add duplicate nodes to the heap, in the worst case approaching O(E). Considering a fully connected graph, the first node would itself add N-1 nodes to the heap, the second node would add N-2 nodes to the heap, and so on. It would be possible to construct edge weights such that all other nodes are processed before the N-2 not processed from the outset would remain on the heap until all other nodes have been processed. This would be possible for each of the other N-1 nodes as well. The total size of the heap, while never fully reaching E, would grow with respect to E, making it O(E). Only `decrease_key` or abandoning the use of a heap (with obvious time complexity repercussions) can guarantee to keep the heap size at O(N).
```

Comment:
I don't understand this sentence "It would be possible to construct edge weights such that all other nodes are processed before the N-2 not processed from the outset would remain on the heap until all other nodes have been processed."

Are there extra words in here? Or am I drowsy from a big lunch

Reply:
Key missing word for the flow: that
```suggestion
However, without the `decrease_key` operation, we cannot update the priority of a node that is already in the heap. This means that we will have to add duplicate nodes to the heap, in the worst case approaching O(E). Considering a fully connected graph, the first node would itself add N-1 nodes to the heap, the second node would add N-2 nodes to the heap, and so on. It would be possible to construct edge weights such that all other nodes that are processed before the N-2 not processed from the outset would remain on the heap until all other nodes have been processed. This would be possible for each of the other N-1 nodes as well. The total size of the heap, while never fully reaching E, would grow with respect to E, making it O(E). Only `decrease_key` or abandoning the use of a heap (with obvious time complexity repercussions) can guarantee to keep the heap size at O(N).
```
But I think a little wordsmithing there is in order. Maybe reorder the clauses a bit... Hmm...

Reply:
Ah, yes the 'that' makes it flow better. But it is a little wordy. Agree about the wordsmithing

Reply:
```suggestion
However, without the `decrease_key` operation, we cannot update the priority of a node that is already in the heap. This means that we will have to add duplicate nodes to the heap, in the worst case approaching O(E). Considering a fully connected graph, the first node would itself add N-1 nodes to the heap, the second node would add N-2 nodes to the heap, and so on. It would be possible to construct edge weights such that as we visit each node, all but one of its edge connections would remain unprocessed (that is, on the heap) until after the entire rest of the graph has been visited. The total size of the heap, while never quite fully reaching E, would grow with respect to E, making it O(E). Only `decrease_key` or abandoning the use of a heap (with obvious time complexity repercussions) can guarantee to keep the heap size at O(N).
```

Reply:
Nice!